File: leetcode/ReverseLinkedListII/reverseBetween.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  ListNode *reverseBetween(ListNode *head, int m, int n) {
    // Start typing your C/C++ solution below
    // DO NOT write int main() function
    if (m == n) return head;
    int idx = 1;
    ListNode *cur = head->next, *end = head;
    if (m == 1) {
      while (idx < n) {
	end->next = cur->next;
	cur->next = head;
	head = cur;
	cur = end->next;
	idx++;
      }
    } else {
      ListNode *pre = head;
      idx = 1;
      while (idx < m - 1) {
	pre = pre->next;
	idx++;
      }
      end = pre->next;
      cur = end->next;
      idx++;
      while (idx < n) {
	end->next = cur->next;
	cur->next = pre->next;
	pre->next = cur;
	cur = end->next;
	idx++;
      }
    }
    return head;
  }
};

File: leetcode/ReverseLinkedListII/reverseBetween.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ListNode fakeHead = new ListNode(0), current = head, prev = fakeHead, last = current;
        fakeHead.next = head;
        n -= m;
        while (--m > 0) {
            prev = current;
            current = current.next;
        }
        last = current;
        current = current.next;
        while (n-- > 0) {
            ListNode next = current.next;
            current.next = prev.next;
            prev.next = current;
            last.next = next;
            current = next;
        }
        return fakeHead.next;
    }
}